elastic/kibana
# [Enterprise Search] Add Overview landing page/plugin

## Summary

This adds an Enterprise Search Overview plugin/landing page for the new Kibana home to link to. 🎉 

Changes:
- Sets up new plugin, added to the sidebar, and linked to from Kibana Home
- Conditionally hides/displays the product cards depending on whether the user has access to the product, and disables the plugin entirely if the user does not have access to either AS or WS
- Sets up new plugin telemetry
- Updates our App Search & Workplace Search breadcrumb trails to link to the new overview plugin from the Enterprise Search breadcrumb (was previously not a link)
- Updates our breadcrumbs and react router link helpers with a new `shouldNotCreateHref` flag - this allows us to create RR paths outside of the plugin basename while still navigating within Kibana/between Kibana plugins.
- [Tech debt] Refactor some strings in `public/plugin.ts` to `common/constants.ts`

Please note that this is an MVP plugin that mimics the current Enterprise Search Home (http://localhost:3002/ent/select) - however, @johnbarrierwilson will eventually customize it at some point.

### Screencaps

![enterprise search overview](https://user-images.githubusercontent.com/549407/92185993-35ec8c80-ee0a-11ea-9c87-10c73e47a94e.png)

#### Dark mode:
![darkmode](https://user-images.githubusercontent.com/549407/92185995-36852300-ee0a-11ea-99c2-41f799072058.png)

#### Side nav:
<img width="724" alt="" src="https://user-images.githubusercontent.com/549407/92186178-d04cd000-ee0a-11ea-945b-2f6873761553.png">

#### Telemetry:
<img width="396" alt="" src="https://user-images.githubusercontent.com/549407/92186200-de025580-ee0a-11ea-9b27-e1009fd155ae.png">

## QA

- Navigation
  - [x] Clicking from the Kibana Home to Enterprise Search now takes you to the Overview plugin, not App Search
  - [x] Clicking Enterprise Search > Overview in the sidebar nav takes you to the plugin
  - [x] Clicking "Launch App Search" takes you to App Search
  - [x] Clicking "Enterprise Search" from App Search's breadcrumb trail takes you to the plugin
  - [x] Clicking "Launch Workplace Search" takes you to Workplace Search
  - [x] Clicking "Enterprise Search" from Workplace Search's breadcrumb trail takes you to the plugin
- Overview UI
  - [x] Works in small phone and tablet dimensions
  - [x] Works in dark mode
  - [x] Looks acceptably similar to http://localhost:3002/ent/select (considering EUI conversions & removing bespoke CSS)
- Telemetry
  - Go to http://localhost:5601/api/stats?extended=true
  - [x] Confirm that a new `enterprise_search` plugin is present, and that it has non-zero numbers for `ui_viewed.overview` and `ui_clicked.app_search` and `ui_clicked.workplace_search` that correspond to the number of times you've clicked said buttons (try clicking, then refreshing the stats page and confirming the count increments).
- Access
  - Setup
    - In Kibana, create a custom role (e.g. `test`) with only Dashboard/Discover permissions
    - In Kibana, create a user (e.g. `test_as_access`) with the above role. In App Search, create a role mapping for that username with an access of admin etc. (or dev)
    - In Kibana, create a user (e.g. `test_ws_access`) with the above role. In Workplace Search, create a role mapping for that username with an access of admin or user
    - In Kibana, create a user (e.g. `test_no_access` with the above role. Do not add any role mappings in either product for this username.
  - Log into the `test_as_access` App Search user and confirm that:
    - [x] Only the Overview and App Search plugins show up in the side bar
    - [x] Only the App Search product card shows up in the Overview plugin
  - Log into the `test_ws_access` Workplace Search user and confirm that:
    - [x] Only the Overview and Workplace Search plugins show up in the side bar
    - [x] Only the Workplace Search product card shows up in the Overview plugin
  - Log into the `test_no_access` user and confirm that:
    - [x] No Enterprise Search plugins show up in the sidebar
    - [x] The Enterprise Search card does not show up at all on the Home catalogue
    - [x] http://localhost:5601/app/enterprise_search/overview correctly 404s

### Checklist


## Release note

Adds a new Enterprise Search overview plugin, which introduces and guides users to the App Search and Workplace Search plugins.
